File: chapter19/src/main/java/main.java
import java.util.HashMap;

class JavaApplication {
    public static void main(String[] args) {
        User user = new User();
        user.setFirstName("Testy");
        user.setLastName("McTester son");

        Address address = new Address(
                "345 None Exiting Avenue NW",
                null,
                "Washington",
                "DC",
                "20016",
                AddressType.Shipping
        );

        UserExtensions.addOrUpdateAddress(user, address);
        LabelPrinter.printLabelFor(user);

        Address.JSONKeys keys = Address.JSONKeys.INSTANCE;

        HashMap<String, Object> addressJSON = new HashMap<>();
        addressJSON.put(keys.streetLine1, address.streetLine1);
        addressJSON.put(keys.streetLine2, address.streetLine2);
        addressJSON.put(keys.city, address.city);
        addressJSON.put(keys.stateOrProvince, address.stateOrProvince);
        addressJSON.put(keys.postalCode, address.postalCode);
        addressJSON.put(keys.country, address.country);
        addressJSON.put(keys.addressType, address.addressType);

        System.out.println("Address JSON:\n" + addressJSON);

        System.out.println("Sample first line of address: " + Address.sampleFirstLine);

        Address canadian = Address.canadianSample(AddressType.Shipping);
        System.out.println(canadian);
    }
}

File: chapter19/src/main/kotlin/Main.kt
fun main() {

    val user = User()

    user.firstName = "Bob"
    user.lastName = "Parker"

    println("User info:\n$user")

    val billingAddress = Address(
        "123 Fake Street",
        "4th Floor",
        "Los Angeles",
        "CA",
        "90291",
        AddressType.Billing
    )

    println("Billing Address:\n$billingAddress")

    user.addOrUpdateAddress(billingAddress)

    val shippingAddress = Address(
        "987 Unreal Driver",
        null,
        "Burbank",
        "CA",
        "91523",
        AddressType.Shipping
    )

    user.addOrUpdateAddress(shippingAddress)

    println("User info after adding addresses:\n$user")

    println("Shipping Label:")
    printLabelFor(user)

    val anotherUser = User()
//    anotherUser.addresses = null
    println("Another User has ${anotherUser.addresses.count()} addresses")

    println("Another User first name: ${anotherUser.firstName ?: "(not set)"}")

    println("Sample First Line: ${Address.sampleFirstLine}")

    println("Sample Canadian Address:\n${Address.canadianSample(AddressType.Billing)}")
}